### PR TITLE
crush: detect and (usually) fix ruleset != rule id

### PR DESCRIPTION
**Pull request by liewegas**
Transition all CRUSH maps to stop using ruleset.  Someday we'll be able to
drop it from CRUSH.

### CODE REVIEW

**Comment by dachary** [comment]
It is intended to be a breaking change ? i.e. existing commands using ruleset must be converted to rule + the output that previous had ruleset as a dictionary key now have rule instead ?

**Comment by liewegas** [comment]
Yes, sort of.  The pool set property has allowed crush_rule instead of crush_ruleset for a long time; we'r ejust finally dropping the crush_ruleset variation.

The other changes should only change the output json, replacing crush_ruleset with crush_rule.

It would be nice to just tell consuming apps to change for luminous, but if we really  need to we could keep them both for another cycle..  i just worry that it won't really help, since there's no big "this interface is deprecated!" warning we can show a user.

**Comment by dachary** [comment]
Ok then :-) Reviewing with that in mind.

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -4837,40 +4857,40 @@ int OSDMonitor::prepare_pool_stripe_width(const unsigned pool_type,
...
       {
-	if (ruleset_name == "") {
-	  //Use default ruleset
-	  *crush_ruleset = osdmap.crush->get_osd_pool_default_crush_replicated_ruleset(g_ceph_context);
-	  if (*crush_ruleset < 0) {
-	    // Errors may happen e.g. if no valid ruleset is available
-	    *ss << "No suitable CRUSH ruleset exists, check "
+	if (rule_name == "") {
+	  //Use default rule
+	  *crush_rule = osdmap.crush->get_osd_pool_default_crush_replicated_ruleset(g_ceph_context);
```
Comment (dachary) [inline src/mon/OSDMonitor.cc]: s/ruleset/rule/

Comment (liewegas) [inline src/mon/OSDMonitor.cc]: My thinking is to leave most of the crush ruleset stuff alone until it's been long enough that we can remove it entirely.  Ceph can be restricted to a 'rule' concept and reject crush maps with rulesets containing more than one rule.

Comment (dachary) [inline src/mon/OSDMonitor.cc]: ack

**Comment by jdurgin** [comment]
related to http://tracker.ceph.com/issues/17138

File: src/osd/OSDMap.cc [hunk 2]
```diff
@@ -2979,7 +2979,7 @@ int OSDMap::build_simple(CephContext *cct, epoch_t e, uuid_d &fsid,
       pools[pool].set_flag(pg_pool_t::FLAG_NOSIZECHANGE);
     pools[pool].size = cct->_conf->osd_pool_default_size;
     pools[pool].min_size = cct->_conf->get_osd_pool_default_min_size();
-    pools[pool].crush_ruleset = default_replicated_ruleset;
+    pools[pool].crush_rule = default_replicated_ruleset;
```
Comment (joscollin) [inline src/osd/OSDMap.cc]: ruleset or rule ?

> int const default_replicated_ruleset = crush->get_osd_pool_default_crush_replicated_ruleset(cct);

Here too.

File: src/osd/OSDMap.cc [hunk 3]
```diff
@@ -3135,14 +3135,10 @@ int OSDMap::build_simple_crush_rulesets(CephContext *cct,
 					const string& root,
 					ostream *ss)
 {
-  int crush_ruleset =
-      crush._get_osd_pool_default_crush_replicated_ruleset(cct, true);
+  int crush_ruleset = crush.get_osd_pool_default_crush_replicated_ruleset(cct);
```
Comment (joscollin) [inline src/osd/OSDMap.cc]: ruleset or rule ?

**Comment by gregsfortytwo** [comment]
http://tracker.ceph.com/issues/17138 Should be closed when this merges, please. :)

**Comment by liewegas** [comment]
<pre>
2017-06-16T04:26:33.600 INFO:tasks.ceph.mgr.x.smithi022.stderr:Error on request:
2017-06-16T04:26:33.601 INFO:tasks.ceph.mgr.x.smithi022.stderr:Traceback (most recent call last):
2017-06-16T04:26:33.601 INFO:tasks.ceph.mgr.x.smithi022.stderr:  File "/usr/lib/python2.7/site-packages/werkzeug/serving.py", line 177, in run_wsgi
2017-06-16T04:26:33.601 INFO:tasks.ceph.mgr.x.smithi022.stderr:    execute(self.server.app)
2017-06-16T04:26:33.601 INFO:tasks.ceph.mgr.x.smithi022.stderr:  File "/usr/lib/python2.7/site-packages/werkzeug/serving.py", line 165, in execute
2017-06-16T04:26:33.601 INFO:tasks.ceph.mgr.x.smithi022.stderr:    application_iter = app(environ, start_response)
2017-06-16T04:26:33.602 INFO:tasks.ceph.mgr.x.smithi022.stderr:  File "/usr/lib/python2.7/site-packages/pecan/middleware/recursive.py", line 56, in __call__
2017-06-16T04:26:33.602 INFO:tasks.ceph.mgr.x.smithi022.stderr:    return self.application(environ, start_response)
2017-06-16T04:26:33.602 INFO:tasks.ceph.mgr.x.smithi022.stderr:  File "/usr/lib/python2.7/site-packages/pecan/core.py", line 570, in __call__
2017-06-16T04:26:33.602 INFO:tasks.ceph.mgr.x.smithi022.stderr:    self.handle_request(req, resp)
2017-06-16T04:26:33.602 INFO:tasks.ceph.mgr.x.smithi022.stderr:  File "/usr/lib/python2.7/site-packages/pecan/core.py", line 508, in handle_request
2017-06-16T04:26:33.602 INFO:tasks.ceph.mgr.x.smithi022.stderr:    result = controller(*args, **kwargs)
2017-06-16T04:26:33.602 INFO:tasks.ceph.mgr.x.smithi022.stderr:  File "/usr/lib64/ceph/mgr/restful/decorators.py", line 33, in decorated
2017-06-16T04:26:33.603 INFO:tasks.ceph.mgr.x.smithi022.stderr:    return f(*args, **kwargs)
2017-06-16T04:26:33.603 INFO:tasks.ceph.mgr.x.smithi022.stderr:  File "/usr/lib64/ceph/mgr/restful/api/osd.py", line 130, in get
2017-06-16T04:26:33.603 INFO:tasks.ceph.mgr.x.smithi022.stderr:    return module.instance.get_osds(pool_id)
2017-06-16T04:26:33.603 INFO:tasks.ceph.mgr.x.smithi022.stderr:  File "/usr/lib64/ceph/mgr/restful/module.py", line 451, in get_osds
2017-06-16T04:26:33.603 INFO:tasks.ceph.mgr.x.smithi022.stderr:    pools_map = self.get_osd_pools()
2017-06-16T04:26:33.604 INFO:tasks.ceph.mgr.x.smithi022.stderr:  File "/usr/lib64/ceph/mgr/restful/module.py", line 422, in get_osd_pools
2017-06-16T04:26:33.604 INFO:tasks.ceph.mgr.x.smithi022.stderr:    for rule in [r for r in crush_rules if r['rule'] == pool['crush_rule']]:
2017-06-16T04:26:33.604 INFO:tasks.ceph.mgr.x.smithi022.stderr:KeyError: 'rule'
</pre>
/a/sage-2017-06-16_02:34:35-rados-wip-sage-testing2-distro-basic-smithi/1292990

**Comment by liewegas** [comment]
<pre>
2017-06-16T05:48:09.108 INFO:tasks.workunit.client.0.smithi059.stdout:GET osd/pool/get.json?pool=rbd&var=crush_rule: json 200
2017-06-16T05:48:09.108 INFO:tasks.workunit.client.0.smithi059.stderr:Traceback (most recent call last):
2017-06-16T05:48:09.108 INFO:tasks.workunit.client.0.smithi059.stderr:  File "/home/ubuntu/cephtest/clone.client.0/qa/workunits/rest/test.py", line 422, in <module>
2017-06-16T05:48:09.108 INFO:tasks.workunit.client.0.smithi059.stderr:    assert(r.myjson['output']['crush_rule'] == 0)
2017-06-16T05:48:09.108 INFO:tasks.workunit.client.0.smithi059.stderr:AssertionError
</pre>
/a/sage-2017-06-16_02:34:35-rados-wip-sage-testing2-distro-basic-smithi/1293239

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-06-20_17:41:31-rados-wip-sage-testing-distro-basic-smithi/

**Pull request merged**